**Merging New Institutional Debt Rates Data**

In [1]:
qui {
forvalues year = 2018/2024 {
copy "https://nces.ed.gov/ipeds/datacenter/data/FLAGS`year'.zip" FLAGS`year'.zip, replace
}
}

In [2]:
use flags2021, clear
*list unitid idx_f
sum idx_f


    Variable |        Obs        Mean    Std. dev.       Min        Max
-------------+---------------------------------------------------------
       idx_f |      6,440    26814.27    88649.92         -2     490285


In [3]:
*qui {
forvalues year=2018/2022 {

unzipfile FLAGS`year'.zip, replace

	insheet using flags`year'.csv, clear

keep unitid idx_f

gen year = `year' +1

local y = `year' + 1

save flags`y', replace
ds
}
*}

    inflating: flags2018.csv

successfully unzipped FLAGS2018.zip to current directory
total processed:  1
        skipped:  0
      extracted:  1
(104 vars, 6,857 obs)
file flags2019.dta saved
unitid  idx_f   year
    inflating: flags2019.csv

successfully unzipped FLAGS2019.zip to current directory
total processed:  1
        skipped:  0
      extracted:  1
(107 vars, 6,559 obs)
file flags2020.dta saved
unitid  idx_f   year
    inflating: flags2020.csv

successfully unzipped FLAGS2020.zip to current directory
total processed:  1
        skipped:  0
      extracted:  1
(111 vars, 6,440 obs)
file flags2021.dta saved
unitid  idx_f   year
    inflating: flags2021.csv

successfully unzipped FLAGS2021.zip to current directory
total processed:  1
        skipped:  0
      extracted:  1
(111 vars, 6,289 obs)
file flags2022.dta saved
unitid  idx_f   year
    inflating: flags2022.csv

successfully unzipped FLAGS2022.zip to current directory
total processed:  1
        skipped:  0
      extract

In [4]:
forvalues year=2023/2023 {

unzipfile FLAGS`year'.zip, replace

	insheet using flags`year'.csv, clear

rename UNITID unitid

keep unitid idx_f

gen year = `year' +1

local y = `year' + 1

save flags`y', replace
}

    inflating: Flags2023.csv

successfully unzipped FLAGS2023.zip to current directory
total processed:  1
        skipped:  0
      extracted:  1
(105 vars, 6,163 obs)
file flags2024.dta saved


In [5]:
forvalues year=2017/2022 {
	copy https://nces.ed.gov/ipeds/datacenter/data/HD`year'_Data_Stata.zip HD`year'_Data_Stata.zip, replace
	unzipfile HD`year'_Data_Stata, replace
      
	insheet using hd`year'_data_stata.csv, clear

    keep unitid control iclevel fips ccbasic instnm hloffer hdegofr1

rename ccbasic c21basic

gen year=`year' + 1

local y = `year' + 1
    save hd`y'a, replace
		}   

    inflating: hd2017_data_stata.csv

successfully unzipped HD2017_Data_Stata.zip to current directory
total processed:  1
        skipped:  0
      extracted:  1
(72 vars, 7,153 obs)
file hd2018a.dta saved
    inflating: hd2018_data_stata.csv

successfully unzipped HD2018_Data_Stata.zip to current directory
total processed:  1
        skipped:  0
      extracted:  1
(73 vars, 6,857 obs)
file hd2019a.dta saved
    inflating: hd2019_data_stata.csv

successfully unzipped HD2019_Data_Stata.zip to current directory
total processed:  1
        skipped:  0
      extracted:  1
(73 vars, 6,559 obs)
file hd2020a.dta saved
    inflating: hd2020_data_stata.csv

successfully unzipped HD2020_Data_Stata.zip to current directory
total processed:  1
        skipped:  0
      extracted:  1
(73 vars, 6,440 obs)
file hd2021a.dta saved
    inflating: hd2021_data_stata.csv

successfully unzipped HD2021_Data_Stata.zip to current directory
total processed:  1
        skipped:  0
      extracted:  1
(74 vars,

In [6]:
qui {

forvalues year=2023/2024 {
	copy https://nces.ed.gov/ipeds/datacenter/data/HD`year'_Data_Stata.zip HD`year'_Data_Stata.zip, replace
	unzipfile HD`year'_Data_Stata, replace
      
	insheet using hd`year'_data_stata.csv, clear

rename UNITID unitid

    keep unitid control iclevel fips c21basic instnm hloffer hdegofr1 

gen year=`year' - 1

local y = `year' - 1
    save hd`y'a, replace
		}   
	}

In [7]:
clear all

qui {
    forvalues year = 2018/2023 {

        * 1. Download zip
        copy "https://nces.ed.gov/ipeds/datacenter/data/EF2021A_Data_Stata.zip" ///
            "EF2021A_Data_Stata.zip", replace

        * 2. Unzip (creates EF2021A_Data_Stata.csv)
        unzipfile "EF2021A_Data_Stata.zip", replace

        * 3. Import the CSV that actually exists
        import delimited "EF2021A_Data_Stata.csv", clear // varnames(1)

        * 4. Standardize variable names to lowercase
        rename *, lower

        * 5. Filter to level 2 (EF, not EFFY)
        * Check once with: describe ef*
        keep if efalev == 2

        * 6. Keep race/ethnicity totals
        keep unitid eftotlt efaiant efasiat efbkaat efhispt ///
             efnhpit efwhitt ef2mort efunknt efnralt

        * 8. Add year and save
        gen year = `year'
        save ef`year', replace
    }
}


In [65]:
*** WRITE CODE TO DOWNLOAD MULTIPLE YEARS OF SFA STUDENT FINANCIAL AID DATA AND MERGE

unitid    xeftotlt  eftotlw   xefaianw  efasiam   xefbkaam  efhispt   xefnhpit  efnhpiw   xefwhitw  ef2morm   xefunknm  efnralt
efalevel  eftotlt   xefaiant  efaianw   xefasiaw  efbkaam   xefhispm  efnhpit   xefwhitt  efwhitw   xef2morw  efunknm   xefnralm
line      xeftotlm  efaiant   xefasiat  efasiaw   xefbkaaw  efhispm   xefnhpim  efwhitt   xef2mort  ef2morw   xefunknw  efnralm
section   eftotlm   xefaianm  efasiat   xefbkaat  efbkaaw   xefhispw  efnhpim   xefwhitm  ef2mort   xefunknt  efunknw   xefnralw
lstudy    xeftotlw  efaianm   xefasiam  efbkaat   xefhispt  efhispw   xefnhpiw  efwhitm   xef2morm  efunknt   xefnralt  efnralw


In [8]:
ds

unitid   efaiant  efbkaat  efnhpit  ef2mort  efnralt
eftotlt  efasiat  efhispt  efwhitt  efunknt  year


In [ ]:
// use unitid gis4n2 gis4n12 gis4n22 gis4n22 gis4n32 gis4n42 gis4n52 ///
//    upgrntn upgrntp upgrntt upgrnta fgrnt_p using sfa2223_data_stata, clear

// merge 1:1 unitid using Data_Single_Year_Policies,

// drop if _merge==1

// save sfa_collections.dta, replace

In [9]:
use hd2023a, clear

merge 1:1 unitid year using ef2023, nogen

merge 1:1 unitid year using flags2023, nogen

forvalues year = 2018 / 2022 {

append using hd`year'a
merge 1:1 unitid year using ef`year', nogen update
merge 1:1 unitid year using flags`year', nogen update

}


    Result                      Number of obs
    -----------------------------------------
    Not matched                           830
        from master                       566  
        from using                        264  

    Matched                             5,506  
    -----------------------------------------

    Result                      Number of obs
    -----------------------------------------
    Not matched                           208
        from master                       144  
        from using                         64  

    Matched                             6,192  
    -----------------------------------------
(variable instnm was str86, now str91 to accommodate using data's values)

    Result                      Number of obs
    -----------------------------------------
    Not matched                         8,383
        from master                     8,083  
        from using                        300  

    Matched                   

In [10]:
save hd_ef_flags18_23, replace

file hd_ef_flags18_23.dta saved


In [11]:
tab year if idx_f~=.


       year |      Freq.     Percent        Cum.
------------+-----------------------------------
       2018 |      6,857       17.47       17.47
       2019 |      6,857       17.47       34.93
       2020 |      6,559       16.71       51.64
       2021 |      6,440       16.40       68.04
       2022 |      6,289       16.02       84.06
       2023 |      6,256       15.94      100.00
------------+-----------------------------------
      Total |     39,258      100.00


In [20]:
insheet using total_offsets/total_offsets_2019_2023.csv, clear

(4 vars, 147 obs)


In [21]:
collapse (min) year, by(district)
list


     +----------------------------------------------+
     |                              district   year |
     |----------------------------------------------|
  1. |          020 - Grossmont-Cuyamaca CCD   2019 |
  2. |                    030 - Imperial CCD   2019 |
  3. |                   050 - MiraCosta CCD   2019 |
  4. |                     060 - Palomar CCD   2019 |
  5. |                090 - Southwestern CCD   2019 |
     |----------------------------------------------|
  6. |                 110 - Butte-Glenn CCD   2019 |
  7. |               120 - Feather River CCD   2019 |
  8. |                    160 - Redwoods CCD   2019 |
  9. | 170 - Shasta-Tehama-Trinity Joint CCD   2019 |
 10. |              180 - Siskiyou Joint CCD   2019 |
     |----------------------------------------------|
 11. |                    230 - Los Rios CCD   2019 |
 12. |                 240 - Napa Valley CCD   2019 |
 13. |               260 - Sonoma County JCD   2019 |
 14. |                310 -

**NOTE: Pick up here 12/05/2025 by continuing to hard code in idx_f into institutions data set**

In [26]:
insheet using total_offsets/total_offsets_2019_2023.csv, clear
gen idx_f=.

(4 vars, 147 obs)
(147 missing values generated)


In [27]:
replace idx_f = 115287 if regexm(district, "020") == 1   // Grossmont-Cuyamaca CCD
replace idx_f = 115861 if regexm(district, "030") == 1   // Imperial CCD
replace idx_f = 118912 if regexm(district, "050") == 1   // MiraCosta CCD
replace idx_f = 120971 if regexm(district, "060") == 1   // Palomar CCD
replace idx_f = 123800 if regexm(district, "090") == 1   // Southwestern CCD

replace idx_f = 110246 if regexm(district, "110") == 1   // Butte-Glenn CCD
replace idx_f = 114433 if regexm(district, "120") == 1   // Feather River CCD
replace idx_f = 121707 if regexm(district, "160") == 1   // Redwoods CCD
replace idx_f = 123299 if regexm(district, "170") == 1   // Shasta-Tehama-Trinity Joint CCD
replace idx_f = 123484 if regexm(district, "180") == 1   // Siskiyou Joint CCD

replace idx_f = 121707 if regexm(district, "230") == 1   // Los Rios CCD
replace idx_f = 119331 if regexm(district, "240") == 1   // Napa Valley CCD
replace idx_f = 123013 if regexm(district, "260") == 1   // Sonoma County JCD (Used Santa Rose Junior College)
replace idx_f = 112817 if regexm(district, "310") == 1   // Contra Costa CCD
replace idx_f = 112190 if regexm(district, "360") == 1   // San Francisco CCD

replace idx_f = 114831 if regexm(district, "420") == 1   // Foothill-De Anza CCD
replace idx_f = 120290 if regexm(district, "430") == 1   // Ohlone CCD
replace idx_f = 114938 if regexm(district, "440") == 1   // Gavilan Joint CCD
replace idx_f = 115393 if regexm(district, "450") == 1   // Hartnell CCD
replace idx_f = 122737 if regexm(district, "470") == 1   // San José-Evergreen CCD

replace idx_f = 111920 if regexm(district, "480") == 1   // Chabot-Las Positas CCD
replace idx_f = 118718 if regexm(district, "530") == 1   // Merced CCD
replace idx_f = 123217 if regexm(district, "560") == 1   // Sequoias CCD
replace idx_f = 123925 if regexm(district, "570") == 1   // State Center CCD
replace idx_f = 448637 if regexm(district, "580") == 1   // West Hills CCD

replace idx_f = 126100 if regexm(district, "590") == 1   // Yosemite CCD
replace idx_f = 109350 if regexm(district, "620") == 1   // Antelope Valley CCD
replace idx_f = 113193 if regexm(district, "640") == 1   // San Luis Obispo County CCD (Used Cuesta College)
replace idx_f = 111461 if regexm(district, "660") == 1   // Santa Clarita CCD
replace idx_f = 125019 if regexm(district, "680") == 1   // Ventura County CCD (Seltect CCD offices over college itself)

replace idx_f = 124113 if regexm(district, "690") == 1   // West Kern CCD (Used Taft)
replace idx_f = 112686 if regexm(district, "710") == 1   // Compton CCD
replace idx_f = 113980 if regexm(district, "720") == 1   // El Camino CCD
replace idx_f = 117681 if regexm(district, "740") == 1   // Los Angeles CCD
replace idx_f = 121044 if regexm(district, "770") == 1   // Pasadena Area CCD (Used Pasadena City College)

replace idx_f = 111887 if regexm(district, "810") == 1   // Cerritos CCD
replace idx_f = 112172 if regexm(district, "820") == 1   // Citrus CCD
replace idx_f = 119164 if regexm(district, "850") == 1   // Mt. San Antonio CCD
replace idx_f = 120023 if regexm(district, "860") == 1   // North Orange County CCD
replace idx_f = 121886 if regexm(district, "880") == 1   // Rio Hondo CCD

replace idx_f = 432144 if regexm(district, "890") == 1   // South Orange County CCD
replace idx_f = 109907 if regexm(district, "910") == 1   // Barstow CCD
replace idx_f = 111939 if regexm(district, "920") == 1   // Chaffey CCD
replace idx_f = 113573 if regexm(district, "930") == 1   // Desert CCD
replace idx_f = 119216 if regexm(district, "940") == 1   // Mt. San Jacinto CCD

replace idx_f = 121901 if regexm(district, "960") == 1   // Riverside CCD
replace idx_f = 428426 if regexm(district, "980") == 1   // San Bernardino CCD  (Used CCD > CCC)
replace idx_f = 125091 if regexm(district, "990") == 1   // Victor Valley CCD


(2 real changes made)
(2 real changes made)
(3 real changes made)
(5 real changes made)
(2 real changes made)
(4 real changes made)
(3 real changes made)
(4 real changes made)
(4 real changes made)
(4 real changes made)
(3 real changes made)
(2 real changes made)
(4 real changes made)
(3 real changes made)
(1 real change made)
(1 real change made)
(3 real changes made)
(1 real change made)
(1 real change made)
(3 real changes made)
(3 real changes made)
(5 real changes made)
(3 real changes made)
(2 real changes made)
(5 real changes made)
(4 real changes made)
(5 real changes made)
(2 real changes made)
(4 real changes made)
(4 real changes made)
(3 real changes made)
(1 real change made)
(4 real changes made)
(3 real changes made)
(5 real changes made)
(5 real changes made)
(4 real changes made)
(3 real changes made)
(5 real changes made)
(2 real changes made)
(3 real changes made)
(1 real change made)
(2 real changes made)
(2 real changes made)
(2 real changes made)
(3 real changes 

In [28]:
save idx_total_offsets_2019_2023, replace

file idx_total_offsets_2019_2023.dta saved


In [29]:
ds

district      count_offset  total_amou~t  year          idx_f


In [18]:
use hd_ef_flags18_23, clear

keep if control==1 & iclevel>1 & year==2023 & fips==6

set line 200

sort instnm

list unitid idx_f instnm

(40,518 observations deleted)

     +-----------------------------------------------------------------------------------+
     | unitid    idx_f                                                            instnm |
     |-----------------------------------------------------------------------------------|
  1. | 491464       -2                                                  ABC Adult School |
  2. | 108807       -2                                             Allan Hancock College |
  3. | 109208   117900                                            American River College |
  4. | 417105       -2                          Baldwin Park Adult & Community Education |
  5. | 109907       -2                                         Barstow Community College |
     |-----------------------------------------------------------------------------------|
  6. | 449357       -2                                             Beaumont Adult School |
  7. | 125170   121178                                     

In [172]:
use hd_ef_flags18_23, clear

*Replaces idx_f obs. with of -2 with value of unitid provided in observation*

replace idx_f=unitid if idx_f==-2

*aggregates the dataset down to one row per idx_f × year, computing different summaries for different variable sets*

collapse (sum) eftotlt-efnralt (first) instnm-c21basic, by(idx_f year)

*keep observations marked as 6, which is California*
keep if fips==6

save hd_ef_flags18_23_ca, replace

(34,951 real changes made)
(32,475 observations deleted)
(file hd_ef_flags18_23_ca.dta not found)
file hd_ef_flags18_23_ca.dta saved


In [175]:
set line 150
gsort -year control -iclevel instnm
list idx_f instnm if control==1 & iclevel>1 & year==2023


      +--------------------------------------------------------------------------+
      |  idx_f                                                            instnm |
      |--------------------------------------------------------------------------|
   1. | 491464                                                  ABC Adult School |
   2. | 417105                          Baldwin Park Adult & Community Education |
   3. | 490212                          Baldy View Regional Occupational Program |
   4. | 449357                                             Beaumont Adult School |
   5. | 487773                        Butte County Regional Occupational Program |
      |--------------------------------------------------------------------------|
   6. | 123493                       Charles A Jones Career and Education Center |
   7. | 112367                                            Clovis Adult Education |
   8. | 492306                                            Escondido Adult School |
   

In [ ]:
clear all

local years 2019 2020 2021 2022 2023

tempfile offsets_all
local first = 1

foreach y of local years {

    import excel using "total_offsets/`y'_total_offset.xlsx", ///
        firstrow clear

    * If needed, standardize names ONCE here
    rename *, lower
    * Adjust these renames to match your actual column names
    * e.g., if you have "district" and "total_amount_offset" already, remove renames
    * rename total_amount_offset total_offset

    gen year = `y'

    keep district count_offset total_amount_offset year

    if `first' {
        save `offsets_all', replace
        local first = 0
    }
    else {
        append using `offsets_all'
        save `offsets_all', replace
    }
}

use `offsets_all', clear
save total_offsets_2019_2023.dta, replace
export delimited using "total_offsets/total_offsets_2019_2023.csv", ///
    replace


In [ ]:
use hd2023a, clear

for years = 2018 / 2022 {

append using hd`year'a
merge 1:1 unitid year using ef`year`a, nogen




In [57]:
clear all

local years 2019 2020 2021 2022 2023

tempfile offsets_all
local first = 1

foreach y of local years {

    import excel using "total_offsets/`y'_total_offset.xlsx", ///
        firstrow clear

    * If needed, standardize names ONCE here
    rename *, lower
    * Adjust these renames to match your actual column names
    * e.g., if you have "district" and "total_amount_offset" already, remove renames
    * rename total_amount_offset total_offset

    gen year = `y'

    keep district count_offset total_amount_offset year

    if `first' {
        save `offsets_all', replace
        local first = 0
    }
    else {
        append using `offsets_all'
        save `offsets_all', replace
    }
}

use `offsets_all', clear
save total_offsets_2019_2023.dta, replace
export delimited using "total_offsets/total_offsets_2019_2023.csv", ///
    replace


(3 vars, 46 obs)
  (all newnames==oldnames)
(file C:\Users\rabid\AppData\Local\Temp\ST_c078_00001b.tmp not found)
file C:\Users\rabid\AppData\Local\Temp\ST_c078_00001b.tmp saved as .dta
    format
(3 vars, 39 obs)
  (all newnames==oldnames)
file C:\Users\rabid\AppData\Local\Temp\ST_c078_00001b.tmp saved as .dta
    format
(3 vars, 32 obs)
  (all newnames==oldnames)
file C:\Users\rabid\AppData\Local\Temp\ST_c078_00001b.tmp saved as .dta
    format
(3 vars, 18 obs)
  (all newnames==oldnames)
(variable district was str41, now str42 to accommodate using data's values)
file C:\Users\rabid\AppData\Local\Temp\ST_c078_00001b.tmp saved as .dta
    format
(3 vars, 12 obs)
  (all newnames==oldnames)
(variable district was str41, now str42 to accommodate using data's values)
file C:\Users\rabid\AppData\Local\Temp\ST_c078_00001b.tmp saved as .dta
    format
file total_offsets_2019_2023.dta saved
file total_offsets/total_offsets_2019_2023.csv saved


In [58]:
ds

district      count_offset  total_amou~t  year


**Sanity Check: Verify Total Number of CCD Observations 147 out of 147**

In [61]:
set line 150
list district count_offset total_amount_offset year


     +-------------------------------------------------------------------------+
     |                                  district   count_~t   total_a~t   year |
     |-------------------------------------------------------------------------|
  1. | 620 - Antelope Valley CCD                        129    19460.63   2023 |
  2. | 110 - Butte-Glenn CCD                            873   172975.36   2023 |
  3. | 810 - Cerritos CCD                              1028   261539.18   2023 |
  4. | 530 - Merced CCD                                1226   255185.95   2023 |
  5. | 860 - North Orange County CCD                    968   198888.81   2023 |
     |-------------------------------------------------------------------------|
  6. | 060 - Palomar CCD                                348   114556.95   2023 |
  7. | 770 - Pasadena Area CCD                          843   241441.77   2023 |
  8. | 160 - Redwoods CCD                                86    19453.90   2023 |
  9. | 960 - Riverside CCD 

In [10]:
* copy "https://nces.ed.gov/ipeds/datacenter/data/FLAGS2023.zip" ///
 * flags2023_data_stata.zip, replace

* unzipfile flags2023_data_stata.zip, replace

(file flags2023_data_stata.zip not found)


In [11]:
* unzipfile flags2023_data_stata.zip, replace

    inflating: Flags2023.csv

successfully unzipped flags2023_data_stata.zip to current directory
total processed:  1
        skipped:  0
      extracted:  1


In [38]:
insheet using Flags2023.csv, clear
save flags2023, replace

(105 vars, 6,163 obs)
file flags2023.dta saved


In [39]:
insheet using debt_thresholds_2024.csv, clear
save debt_thresholds_2024, replace

(10 vars, 35 obs)
file debt_thresholds_2024.dta saved


In [50]:
use flags2023, clear
describe


Contains data from flags2023.dta
 Observations:         6,163                  
    Variables:           105                  5 Dec 2025 13:05
-------------------------------------------------------------------------------
Variable      Storage   Display    Value
    name         type    format    label      Variable label
-------------------------------------------------------------------------------
UNITID          long    %12.0g                UNITID
stat_ic         byte    %8.0g                 STAT_IC
lock_ic         byte    %8.0g                 LOCK_IC
imp_ic          byte    %8.0g                 IMP_IC
stat_c          byte    %8.0g                 STAT_C
lock_c          byte    %8.0g                 LOCK_C
prch_c          byte    %8.0g                 PRCH_C
idx_c           long    %12.0g                IDX_C
pcc_f           byte    %8.0g                 PCC_F
imp_c           byte    %8.0g                 IMP_C
stat_e12        byte    %8.0g                 STAT_E12
lock_e12 

In [52]:
use flags2023, clear

* Make variable names consistent with other file
rename UNITID unitid

* Keep only the two needed variables from flags2023
keep unitid idx_f
describe


Contains data from flags2023.dta
 Observations:         6,163                  
    Variables:             2                  5 Dec 2025 13:05
-------------------------------------------------------------------------------
Variable      Storage   Display    Value
    name         type    format    label      Variable label
-------------------------------------------------------------------------------
unitid          long    %12.0g                UNITID
idx_f           long    %12.0g                IDX_F
-------------------------------------------------------------------------------
Sorted by: 
     Note: Dataset has changed since last saved.


In [54]:
merge 1:1 unitid using debt_thresholds_2024

tab _merge
drop if _merge == 2   // optional: drop institutions that only appear in thresholds
drop _merge

save districtids_debt_thresholds_2024, replace


    Result                      Number of obs
    -----------------------------------------
    Not matched                         6,128
        from master                     6,128  (_merge==1)
        from using                          0  (_merge==2)

    Matched                                35  (_merge==3)
    -----------------------------------------

   Matching result from |
                  merge |      Freq.     Percent        Cum.
------------------------+-----------------------------------
        Master only (1) |      6,128       99.43       99.43
            Matched (3) |         35        0.57      100.00
------------------------+-----------------------------------
                  Total |      6,163      100.00
(0 observations deleted)
(file districtids_debt_thresholds_2024.dta not found)
file districtids_debt_thresholds_2024.dta saved


In [56]:
ds

unitid        school_name   threshold_~e  ccd_discha~s
idx_f         threshold_~l  canceled_d~s  ccd_total_~d
segment       threshold_~r  v8


In [ ]:
use 

In [ ]:
merge 1:1 unitid using debt_thresholds_2024

tab _merge
drop if _merge == 2   // optional: drop institutions that only appear in thresholds
drop _merge

save districtids_debt_thresholds_2024, replace

In [100]:
insheet using ///
"C:\Users\rabid\Graduate Research\Updated_CA_Institutional_Debts\data\Data_Single_Year_Policies.csv", clear

(13 vars, 141 obs)


In [101]:
save Data_Single_Year_Policies, replace

file Data_Single_Year_Policies.dta saved


In [102]:
* copy "https://nces.ed.gov/ipeds/datacenter/data/SFA2223_Data_Stata.zip" ///
 * SFA2223_Data_Stata.zip, replace

In [103]:
* unzipfile SFA2223_Data_Stata.zip, replace

In [104]:
insheet using sfa2223_data_stata.csv, clear
save sfa2223_data_stata, replace

(691 vars, 5,653 obs)
file sfa2223_data_stata.dta saved


In [105]:
use Data_Single_Year_Policies, clear

merge 1:1 unitid using sfa2223_data_stata

drop if _merge == 2   // drop observations only in using

keep unitid gis4n2 gis4n12 gis4n22 gis4n32 gis4n42 gis4n52 ///
     upgrntn upgrntp upgrntt upgrnta fgrnt_p fgrnt_a ///
     school_name year explicit_transcript_holds ///
     ambiguous_transcript_holds explicit_enrollment_holds ///
     ambiguous_enrollment_holds no_info_provided ftb segment

save sfa_collections, replace


    Result                      Number of obs
    -----------------------------------------
    Not matched                         5,520
        from master                         4  (_merge==1)
        from using                      5,516  (_merge==2)

    Matched                               137  (_merge==3)
    -----------------------------------------
(5,516 observations deleted)
file sfa_collections.dta saved


In [97]:
// use unitid gis4n2 gis4n12 gis4n22 gis4n22 gis4n32 gis4n42 gis4n52 ///
//    upgrntn upgrntp upgrntt upgrnta fgrnt_p using sfa2223_data_stata, clear

// merge 1:1 unitid using Data_Single_Year_Policies,

// drop if _merge==1

// save sfa_collections.dta, replace

In [98]:
// list unitid year ftb if _n < 11 //

In [108]:
import delimited "C:\Users\rabid\Graduate Research\Updated_CA_Institutional_Debts\data\FTB_2023.csv", clear
save FTB_2023,replace

(encoding automatically selected: UTF-8)
(9 vars, 75 obs)
file FTB_2023.dta saved


In [109]:
ds

segment     agency      pit_amount  year        v9
unitid      pit_count   all_count   all_amount


In [110]:
drop if missing(segment)

(19 observations deleted)


In [111]:
duplicates report unitid


Duplicates in terms of unitid

--------------------------------------
   Copies | Observations       Surplus
----------+---------------------------
        1 |           41             0
        2 |           12             6
        3 |            3             2
--------------------------------------


In [112]:
duplicates list unitid


Duplicates in terms of unitid

  +----------------------+
  | Group   Obs   unitid |
  |----------------------|
  |     1    28   110501 |
  |     1    29   110501 |
  |     1    37   110501 |
  |     2    18   110556 |
  |     2    38   110556 |
  |----------------------|
  |     3    33   110608 |
  |     3    40   110608 |
  |     4     6   110662 |
  |     4     9   110662 |
  |     5     1   110714 |
  |----------------------|
  |     5     4   110714 |
  |     6    12   122597 |
  |     6    24   122597 |
  |     7    21   200002 |
  |     7    23   200002 |
  +----------------------+


In [113]:
bysort unitid: gen nobs = _N
list unitid year agency all_amount if nobs>1, sepby(unitid)


     +-------------------------------------------------------------------+
     | unitid   year                               agency     all_amount |
     |-------------------------------------------------------------------|
  5. | 110501   2023              CSU Chancellor's Office       5,532.00 |
  6. | 110501   2023              CSU CHANCELLOR'S OFFICE     $9,926.17  |
  7. | 110501   2023              CSU CHANCELLOR'S OFFICE    $71,545.02  |
     |-------------------------------------------------------------------|
 11. | 110556   2023                         Fresno State       6,982.94 |
 12. | 110556   2023   CALIFORNIA STATE UNIVERSITY,FRESNO   $140,059.39  |
     |-------------------------------------------------------------------|
 16. | 110608   2023                       CSU Northridge      21,258.79 |
 17. | 110608   2023                       CSU Northridge   $563,062.93  |
     |-------------------------------------------------------------------|
 22. | 110662   2023    

In [114]:
describe


Contains data from FTB_2023.dta
 Observations:            56                  
    Variables:            10                  3 Dec 2025 16:22
-------------------------------------------------------------------------------
Variable      Storage   Display    Value
    name         type    format    label      Variable label
-------------------------------------------------------------------------------
segment         byte    %8.0g                 
unitid          long    %12.0g                
agency          str35   %35s                  
pit_count       str6    %9s                   
pit_amount      str14   %14s                  
all_count       str6    %9s                   
year            int     %8.0g                 
all_amount      str14   %14s                  
v9              byte    %8.0g                 
nobs            float   %9.0g                 
-------------------------------------------------------------------------------
Sorted by: unitid
     Note: Dataset has chan

In [115]:
rename pit_amount  pit_amount_str
rename all_amount  all_amount_str

In [116]:
destring pit_amount_str, gen(pit_amount) ignore("$,")
destring all_amount_str, gen(all_amount) ignore("$,")

pit_amount_str: characters $ , removed; pit_amount generated as double
all_amount_str: characters $ , removed; all_amount generated as double


In [117]:
describe pit_amount all_amount
list unitid year pit_amount all_amount in 1/10


Variable      Storage   Display    Value
    name         type    format    label      Variable label
-------------------------------------------------------------------------------
pit_amount      double  %10.0g                
all_amount      double  %10.0g                

     +---------------------------------------+
     | unitid   year   pit_amo~t   all_amo~t |
     |---------------------------------------|
  1. | 109819   2023     4282.37     4282.37 |
  2. | 110422   2023    95666.74    97968.56 |
  3. | 110486   2023   376633.69    382602.1 |
  4. | 110495   2023   328464.74   330004.13 |
  5. | 110501   2023        5532        5532 |
     |---------------------------------------|
  6. | 110501   2023     9901.32     9926.17 |
  7. | 110501   2023    71538.71    71545.02 |
  8. | 110510   2023    25664.34    25814.02 |
  9. | 110538   2023   211793.43   213404.18 |
 10. | 110547   2023    11867.05    12043.05 |
     +---------------------------------------+


*Original Destring Practice that worked*

In [118]:
*destring pit_count pit_amount all_count all_amount, replace ignore(",") force*

In [119]:
describe pit_count pit_amount all_count all_amount


Variable      Storage   Display    Value
    name         type    format    label      Variable label
-------------------------------------------------------------------------------
pit_count       str6    %9s                   
pit_amount      double  %10.0g                
all_count       str6    %9s                   
all_amount      double  %10.0g                


**Attempting to de-string and then collapes matching unitid observations**

In [120]:
destring pit_count all_count, replace ignore(",")

pit_count: character , removed; replaced as int
all_count: character , removed; replaced as int


In [121]:
bysort unitid: egen total_all_amount = total(all_amount)
bysort unitid: egen total_pit_amount = total(pit_amount)
bysort unitid: egen total_all_count  = total(all_count)
bysort unitid: egen total_pit_count  = total(pit_count)

In [122]:
bysort unitid: keep if _n == 1

(8 observations deleted)


In [123]:
drop all_amount pit_amount all_count pit_count
rename total_all_amount all_amount
rename total_pit_amount pit_amount
rename total_all_count  all_count
rename total_pit_count  pit_count

In [124]:
duplicates report unitid


Duplicates in terms of unitid

--------------------------------------
   Copies | Observations       Surplus
----------+---------------------------
        1 |           48             0
--------------------------------------


In [125]:
. tabstat all_amount, statistics(sum) format(%12.0fc)


    Variable |          Sum
-------------+-------------
  all_amount |   11,987,864
---------------------------


In [126]:
describe pit_count pit_amount all_count all_amount


Variable      Storage   Display    Value
    name         type    format    label      Variable label
-------------------------------------------------------------------------------
pit_count       float   %9.0g                 
pit_amount      float   %9.0g                 
all_count       float   %9.0g                 
all_amount      float   %9.0g                 


In [127]:
duplicates report unitid


Duplicates in terms of unitid

--------------------------------------
   Copies | Observations       Surplus
----------+---------------------------
        1 |           48             0
--------------------------------------


In [128]:
save FTB_2023, replace

file FTB_2023.dta saved


**New Code: Reviewing all_amount totals**

In [132]:
use FTB_2023, clear
merge 1:1 unitid using sfa_collections, 

keep unitid gis4n2 gis4n12 gis4n22 gis4n22 gis4n32 gis4n42 gis4n52 ///
    upgrntn upgrntp upgrntt upgrnta fgrnt_p fgrnt_a school_name year explicit_transcript_holds ///
    ambiguous_transcript_holds explicit_enrollment_holds ambiguous_enrollment_holds ///
    no_info_provided ftb segment pit_count pit_amount all_count all_amount year


    Result                      Number of obs
    -----------------------------------------
    Not matched                           115
        from master                        11  (_merge==1)
        from using                        104  (_merge==2)

    Matched                                37  (_merge==3)
    -----------------------------------------


In [133]:
/*
use FTB_2023, clear
merge 1:1 unitid using sfa_collections, 

keep unitid gis4n2 gis4n12 gis4n22 gis4n22 gis4n32 gis4n42 gis4n52 ///
    upgrntn upgrntp upgrntt upgrnta fgrnt_p school_name year explicit_transcript_holds ///
    ambiguous_transcript_holds explicit_enrollment_holds ambiguous_enrollment_holds ///
    no_info_provided ftb frst_yr_sought last_yr_sought amount_sought segment ///
    pit_count pit_amount all_count all_amount
*/

In [134]:
count

  152


In [135]:
describe


Contains data from FTB_2023.dta
 Observations:           152                  
    Variables:            26                  3 Dec 2025 16:26
-------------------------------------------------------------------------------
Variable      Storage   Display    Value
    name         type    format    label      Variable label
-------------------------------------------------------------------------------
segment         byte    %8.0g                 
unitid          long    %12.0g                
year            int     %8.0g                 
all_amount      float   %9.0g                 
pit_amount      float   %9.0g                 
all_count       float   %9.0g                 
pit_count       float   %9.0g                 
school_name     str35   %35s                  
explicit_tran~s str1    %9s                   
ambiguous_tra~s str1    %9s                   
explicit_enro~s str1    %9s                   
ambiguous_enr~s str1    %9s                   
no_info_provi~d str1    %9s    

**Checking for Merge Changes /w all_amount from FTB_2023 data set**

In [136]:
. tabstat all_amount, statistics(sum) format(%12.0fc)


    Variable |          Sum
-------------+-------------
  all_amount |   11,987,864
---------------------------


In [138]:
save sfa_collections, replace

file sfa_collections.dta saved


**Pull data from effy data sets online** [LEFT OFF HERE]

In [152]:
clear all

qui {
    forvalues year = 2021/2021 {

        * 1. Download zip
        copy "https://nces.ed.gov/ipeds/datacenter/data/EF2021A_Data_Stata.zip" ///
            "EF2021A_Data_Stata.zip", replace

        * 2. Unzip (creates EF2021A_Data_Stata.csv)
        unzipfile "EF2021A_Data_Stata.zip", replace

        * 3. Import the CSV that actually exists
        import delimited "EF2021A_Data_Stata.csv", clear varnames(1)

        * 4. Standardize variable names to lowercase
        rename *, lower

        * 5. Filter to level 2 (EF, not EFFY)
        * Check once with: describe ef*
        keep if efalev == 2

        * 6. Keep race/ethnicity totals
        keep unitid efytotlt efyaiant efyasiat efybkaat efyhispt ///
             efynhpit efywhitt efy2mort efyunknt efynralt

        * 7. Rename to ttl*
        foreach var in efytotlt efyaiant efyasiat efybkaat efyhispt ///
                       efynhpit efywhitt efy2mort efyunknt efynralt {
            rename `var' ttl`var'
        }

        * 8. Add year and save
        gen year = `year'
        save ttlef`year', replace
    }
}


file not found
    ef2021a_data_stata.csv (The process cannot access the file because it is be
> ing used by another process)
r(601);


**Old Code That Worked, aside from the first set that I'm currently working on (12/03/2025)**

In [149]:

qui {
forvalues year=2021/2021 {
    copy https://nces.ed.gov/ipeds/datacenter/data/EF2021A_Data_Stata.zip ef`year'_Data_Stata.zip, replace
    unzipfile ef`year'_Data_Stata.zip, replace
     
    insheet using ef`year'a_data_stata.csv, clear

    keep if efalev==2

    keep UNITID efytotlt efyaiant efyasiat efybkaat efyhispt efynhpit efywhitt efy2mort efyunknt efynralt

    foreach var in efytotlt efyaiant efyasiat efybkaat efyhispt efynhpit efywhitt efy2mort efyunknt efynralt {
    rename `var' ttl`var'
    }
ds
gen year=`year'
    save ttlef`year', replace
        }
    }


file not found
    ef2021a_data_stata.csv (The process cannot access the file because it is be
> ing used by another process)
r(601);


In [141]:

qui {
forvalues year=2023/2023 {
    copy https://nces.ed.gov/ipeds/datacenter/data/EFFY2023_Data_Stata.zip effy`year'_Data_Stata.zip, replace
    unzipfile effy`year'_Data_Stata.zip, replace
     
    insheet using effy`year'_data_stata.csv, clear

    keep if effyalev==22

    keep UNITID efytotlt efyaiant efyasiat efybkaat efyhispt efynhpit efywhitt efy2mort efyunknt efynralt

    foreach var in efytotlt efyaiant efyasiat efybkaat efyhispt efynhpit efywhitt efy2mort efyunknt efynralt {
    rename `var' ft`var'
    }
ds
gen year=`year'
    save fteffy`year', replace
        }
    }


In [ ]:

qui {
forvalues year=2023/2023 {
    copy https://nces.ed.gov/ipeds/datacenter/data/EFFY2023_Data_Stata.zip effy`year'_Data_Stata.zip, replace
    unzipfile effy`year'_Data_Stata.zip, replace
     
    insheet using effy`year'_data_stata.csv, clear

    keep if effyalev==42

    keep unitid efytotlt efyaiant efyasiat efybkaat efyhispt efynhpit efywhitt efy2mort efyunknt efynralt

    foreach var in efytotlt efyaiant efyasiat efybkaat efyhispt efynhpit efywhitt efy2mort efyunknt efynralt {
    rename `var' pt`var'
    }
ds
gen year=`year'
    save pteffy`year', replace
        }
    }


**HD Data Set Pull**

In [ ]:
qui {
forvalues year=2023/2023 {
	copy https://nces.ed.gov/ipeds/datacenter/data/HD`year'_Data_Stata.zip HD`year'_Data_Stata.zip, replace
	unzipfile HD`year'_Data_Stata, replace
      
	insheet using hd`year'_data_stata.csv, clear
   
    keep unitid control iclevel fips ccbasic instnm hloffer hdegofr1
    save hd`year'a, replace
		}   
	}

In [ ]:
describe, fullname

In [ ]:
/*
use ttleffy2023, clear
duplicates report unitid
*/

In [ ]:
use hd2023a, clear

merge 1:1 unitid using ttleffy2023, nogen
merge 1:1 unitid using fteffy2023, nogen
merge 1:1 unitid using pteffy2023, nogen

In [ ]:
save hd_effy_2023, replace

In [ ]:
use hd_effy_2023.dta

**Original Merge Code for Combining sfa_collections /w master hd_effy_2023 data set**

In [ ]:
/*
use sfa_collections, clear
merge 1:1 unitid using hd_effy_2023, gen(collect_hd_merge)

drop if collect_hd_merge!=3
save final_merged_data_set, replace
*/

*New Attempt at Merging ~/w Loss of 3 obs**

In [ ]:
use sfa_collections, clear
merge 1:1 unitid using hd_effy_2023, gen(collect_hd_merge) keep(master match)
save final_merged_data_set, replace

In [ ]:
use final_merged_data_set
ds

In [ ]:
. tabstat all_amount, statistics(sum) format(%12.0fc)

# Troubling Shooting New Merged Final Data Set: Sanity Check on Other Variables

In [ ]:
use final_merged_data_set, clear

In [ ]:
ds

In [ ]:
. tabstat all_amount, statistics(sum) format(%12.0fc)

In [ ]:
misstable summarize _all

In [ ]:
ds, has(type numeric)
local nums `r(varlist)'
summarize `nums', detail

In [ ]:
quietly summarize pit_amount 
display "Total pit_amount = $" string(r(sum),"%12.0fc")
quietly summarize all_count 
display "Total all_count   = " r(sum)

In [ ]:
tabulate segment, missing nofreq
tabulate explicit_transcript_holds, missing
tabulate ambiguous_transcript_holds, missing
tabulate ftb, missing

In [ ]:
list unitid pit_amount if pit_amount<0
list unitid all_count   if all_count   <0

In [ ]:
duplicates report unitid

In [ ]:
egen sum_by_race = rowtotal(ttlefytotlt*)  // whatever race vars you have
generate diff = sum_by_race - ttlefytotlt
assert diff==0

In [ ]:
list unitid ttlefytotlt sum_by_race diff if diff!=0

In [ ]:
tabstat all_amount pit_amount all_count, by(segment) ///
    statistics(N sum mean sd) format(%12.0fc %12.0fc %9.0g %9.2f)

# Trouble shooting the all_amount drop off that occurs as a result of data merges

In [ ]:
use FTB_2023.dta, clear
quietly summarize all_amount
local raw_N  = r(N)
local raw_sum= r(sum)

In [ ]:
use final_merged_data_set.dta, clear
quietly summarize all_amount
local mrg_N  = r(N)
local mrg_sum= r(sum)

In [ ]:
di as txt "RAW:     obs=" as res `raw_N'  "   sum=$" as res string(`raw_sum',"%12.0fc")
di as txt "MERGED:  obs=" as res `mrg_N'  "   sum=$" as res string(`mrg_sum',"%12.0fc")

In [ ]:
use FTB_2023.dta, clear

In [ ]:
merge 1:1 unitid using final_merged_data_set.dta

In [ ]:
tabulate _merge

In [ ]:
list unitid all_amount segment if _merge==1

In [ ]:
quietly summarize all_amount if _merge==1
display as text "Dropped OBS = " as result r(N) ///
       "   Dropped SUM = $" as result string(r(sum),"%12.0fc")

**Revieiwng the different merge locations to figure out where the observations are getting dropped**

In [ ]:
use FTB_2023.dta, clear
quietly summarize all_amount
local before_obs = r(N)
local before_sum = r(sum)
di as txt "🔹 RAW:   obs=`before_obs'   sum=$" as result string(`before_sum',"%12.0fc")

In [ ]:
merge 1:1 unitid using sfa_collections.dta
tabulate _merge
quietly summarize all_amount if _merge==3
local after1_obs = r(N)
local after1_sum = r(sum)
di as txt "🔹 After 1st merge:  obs=`after1_obs'   sum=$" as result string(`after1_sum',"%12.0fc")


In [ ]:
/*
use Data_Single_Year_Policies, clear
merge 1:1 unitid using sfa2223_data_stata, 

 drop if _merge==2

keep unitid gis4n2 gis4n12 gis4n22 gis4n22 gis4n32 gis4n42 gis4n52 ///
    upgrntn upgrntp upgrntt upgrnta fgrnt_p fgrnt_a school_name year explicit_transcript_holds ///
    ambiguous_transcript_holds explicit_enrollment_holds ambiguous_enrollment_holds ///
    no_info_provided ftb segment

save sfa_collections, replace
*/

In [ ]:
keep if _merge==3
drop _merge
merge 1:1 unitid year using sfa_collections.dta
tabulate _merge
quietly summarize all_amount if _merge==3
local after2_obs = r(N)
local after2_sum = r(sum)
di as txt "🔹 After 2nd merge:  obs=`after2_obs'   sum=$" as result string(`after2_sum',"%12.0fc")

In [ ]:
/*
use FTB_2023, clear
merge 1:1 unitid using sfa_collections, 

keep unitid gis4n2 gis4n12 gis4n22 gis4n22 gis4n32 gis4n42 gis4n52 ///
    upgrntn upgrntp upgrntt upgrnta fgrnt_p fgrnt_a school_name year explicit_transcript_holds ///
    ambiguous_transcript_holds explicit_enrollment_holds ambiguous_enrollment_holds ///
    no_info_provided ftb segment pit_count pit_amount all_count all_amount year
*/

In [ ]:
keep if _merge==3
drop _merge
merge 1:1 unitid using third_dataset.dta
tabulate _merge
quietly summarize all_amount if _merge==3
local after3_obs = r(N)
local after3_sum = r(sum)
di as txt "🔹 After 3rd merge:  obs=`after3_obs'   sum=$" as result string(`after3_sum',"%12.0fc")

In [ ]:
/*
use sfa_collections, clear
merge 1:1 unitid using hd_effy_2023, gen(collect_hd_merge)

drop if collect_hd_merge!=3
save final_merged_data_set, replace
*/